metricproviders/datadog/datadog.go

```diff
@@ -0,0 +1,172 @@
+package datadog
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/evaluate"
+	metricutil "github.com/argoproj/argo-rollouts/utils/metric"
+	templateutil "github.com/argoproj/argo-rollouts/utils/template"
+	log "github.com/sirupsen/logrus"
+	dd "github.com/zorkian/go-datadog-api"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	// ProviderType indicates the provider is datadog
+	ProviderType = "Datadog"
+	// DatadogSecretName is a k8s secret that holds datadog api and app keys
+	DatadogSecretName = "datadog-api-keys"
+)
+
+// Provider contains all the required components to run a datadog query
+type Provider struct {
+	client dd.Client
+	logCtx log.Entry
+}
+
+// Type incidates provider is a datadog provider
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+// Run queries DataDog for a metric
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+	newMeasurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+
+	query, err := templateutil.ResolveArgs(metric.Provider.Datadog.Query, run.Spec.Args)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	// TODO (cabrinha) make from and to configurable
+	from := time.Now().Unix() - 60
+	to := time.Now().Unix()
+	response, err := p.client.QueryMetrics(from, to, query)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	// If datadog returns an empty []Series, retry with a longer window
+	var retries int
+	for retries = 0; retries < 3 && len(response) < 1; retries++ {
+		response, err = p.client.QueryMetrics(from-60, to, query)
+		if err != nil {
+			return metricutil.MarkMeasurementError(newMeasurement, err)
+		}
+
+		if len(response) < 1 {
+			time.Sleep(1 * time.Second)
+		}
+	}
```

Comment:
Is it possible to have a query result change after waiting some time?

Reply:
If so, we should mark this measurement as not finished and have it retry in the resume function instead of holding this goroutine.

Reply:
A query result won't change for a given from and to time. If the from and to change, the result could.

Reply:
I've moved this logic into the Resume() func and reused the Run() func there -- I thought that repeating all the code thats in Run() wasn't the right way to go.